xtrack/madng_interface.py

```diff
@@ -391,11 +515,481 @@ def prepare(self, force=False):
+            dpy    = {init['dpy'+quantity_appendix,start_loc] if madng_init_flag else init['dpy',start_loc]},
+        }}"""
+
+        rmat_array_str = ''
+        if self.match_rmat:
+            rmat_array_str = f"{XSUITE_MADNG_ENV_NAME}.rmat_map_arr = table.new({self.sum_rmat_tar}, 0)\n"
+
+        mng_init_str = r'''
+            ''' + XSUITE_MADNG_ENV_NAME + r''' = {} -- to avoid variable name clashes
+            local obs_flag = MAD.element.flags.observed
+
+            local pts=''' + observables_str + r'''
+
+            ''' + self.mng._sequence_name + r''':select(obs_flag, {list=pts})
+
+            local params = ''' + param_list_str + r'''
+
+            local X0 = MAD.damap {
+                nv=6, -- number of variables
+                mo=2, -- max order of variables
+                np=#params, -- number of parameters
+                po=1, -- max order of parameters
+                pn=params, -- parameter names
+            }
+
+            -- Converting to TPSA (mutating type)
+            for _, v in ipairs(params) do
+                MADX[v] = MADX[v] + X0[v]
+            end
+
+            ''' + init_cond_str + r'''
+
+            local map1 = MAD.gphys.bet2map(B0, X0:copy())
+
+            ''' + coord_assign + r'''
+
+            -- Maps target locations to damaps
+            -- e.g. { 'BPM1' = damap1, 'BPM2' = damap2, ... }
+            ''' + XSUITE_MADNG_ENV_NAME + r'''.target_loc_map = table.new(0, ''' + str(len(self.optics_target_locations)) + r''')
+
+            -- Maps rmat tags to rmat damaps
+            -- e.g. {rmat_damap1, rmat_damap2, ... }
+            ''' + rmat_array_str + r'''
+
+            -- Array of targets with additional info (location, quantity, orbit/optical function)
+            -- e.g. { { loc = 'BPM1', qty = 'beta11', optfun = true }, { loc = 'BPM2', qty = 'x', orbit = 1 }, ... }
+            ''' + XSUITE_MADNG_ENV_NAME + r'''.targets_arr = table.new(''' + str(len(self.targets)) + r''', 0)
+
+            -- List (Array) of target optics/orbit quantities (suitable for MAD-NG),
+            -- e.g. {'beta11', 'alfa11', ...}
+            ''' + XSUITE_MADNG_ENV_NAME + r'''.tar_optics_qtys = ''' + optics_qty_str + r'''
+
+            -- Initial map for tracking/twiss
+            ''' + XSUITE_MADNG_ENV_NAME + r'''.init_X0_map = map1
+
+            -- Identity map
+            ''' + XSUITE_MADNG_ENV_NAME + r'''.empty_X0 = X0
+
+            -- Defining targets array
+            ''' + targets_map_str + r'''
+
+            -- Mapping from xsuite quantity names to madng quantity names
+            ''' + xs_ng_target_map + r'''
+            '''
+
+
+        self.mng.send(mng_init_str)
+
+        self._already_prepared = True
+
+    def run(self):
+        '''
+        Execute the MAD-NG TPSA matching action.
+        This method performs either a Twiss or Track operation in MAD-NG
+        depending if quantities can be calculated through tracking or not.
```

Comment:
See my later comment of the `twiss_flag`. Technically, here you "perform twiss or track operation in MAD-NG depending" on the `self.twiss_flag`, the logic for the quantities is external to the class, and as far as the action is concerned it doesn't matter. (My later comment: maybe this logic could be pulled in? Not sure if it makes sense, you have the big picture so you will know best.)